### PR TITLE
support install rancher-alerting-drivers in hardened cluster

### PR DESCRIPTION
**Pull request by jiaqiluo**
Issues:
https://github.com/rancher/rancher/issues/32358
https://github.com/rancher/rancher/issues/32665

Notes about the PR:
- The chart is tested on a custom hardened cluster under the [CIS 1.6 Benchmark](https://rancher.com/docs/rancher/v2.x/en/security/rancher-2.5/1.6-hardening-2.5/). 
- The app is installed into the hardened cluster successfully.
- After the app is installed, pass the CIS scan with the profile `rke-profile-hardened-1.6`

### CODE REVIEW

**Review by nickgerace — APPROVED** [review]
Left a nit, but LGTM

File: packages/rancher-alerting-drivers/charts/templates/_helpers.tpl
Comment (jiaqiluo) [reply packages/rancher-alerting-drivers/charts/templates/_helpers.tpl:17]: We should be fine because pods will still be assigned to nodes if the target taint does not exist. 
Those charts are curated for Rancher's clusters, so we don't have to put them in values.yaml for now. If we decide to move then we need to do it for all charts :)

File: packages/rancher-alerting-drivers/charts/templates/network-policy-allow-all.yaml
Comment (aiyengar2) [inline packages/rancher-alerting-drivers/charts/templates/network-policy-allow-all.yaml:1]: Instead of adding the network policy, serviceaccount, psp, and roles manually in these files, let's just copy what we have for Monitoring today here:

https://github.com/rancher/charts/blob/dev-v2.5-source/packages/rancher-monitoring/generated-changes/overlay/templates/rancher-monitoring/hardened.yaml

That way we can standardize this across feature charts, similar to what we do today for helpers.tpl functions.

Comment (aiyengar2) [reply packages/rancher-alerting-drivers/charts/templates/network-policy-allow-all.yaml:1]: You still need the serviceaccount, rolebinding, and PSP to allow your deployments to be deployed in a hardened cluster, but this ^^ takes care of the scan requirements.

**Review by aiyengar2 — CHANGES_REQUESTED** [review]

File: packages/rancher-alerting-drivers/charts/templates/network-policy-allow-all.yaml
Comment (jiaqiluo) [reply packages/rancher-alerting-drivers/charts/templates/network-policy-allow-all.yaml]: changes are made based on the above comments.

**Review by nickgerace — APPROVED** [review]
LGTM after Arvind's changes

File: packages/rancher-alerting-drivers/package.yaml
Comment (cbron) [inline packages/rancher-alerting-drivers/package.yaml:2]: Nit: 01

File: packages/rancher-alerting-drivers/charts/templates/hardened.yaml
Comment (cbron) [inline packages/rancher-alerting-drivers/charts/templates/hardened.yaml:25]: Is this the only way to make hardened cluster pass? By changing the default service account ? I haven’t seen this pattern before. cc @aiyengar2

File: packages/rancher-alerting-drivers/package.yaml
Comment (jiaqiluo) [reply packages/rancher-alerting-drivers/package.yaml]: resoved.

File: packages/rancher-alerting-drivers/charts/templates/hardened.yaml
Comment (aiyengar2) [reply packages/rancher-alerting-drivers/charts/templates/hardened.yaml:25]: @cbron you need to patch the default service account in any namespace that you create as part of the chart to be un-automounted. Declared as a req in https://rancher.com/docs/rancher/v2.x/en/security/rancher-2.5/1.6-hardening-2.5/#set-automountserviceaccounttoken-to-false-for-default-service-accounts.

**Comment by aiyengar2** [comment]
@jiaqiluo the CI issue is due to an unrelated problem tracked inhttps://github.com/rancher/charts/pull/1190. Thanks for filing the issue!

**Review by aiyengar2 — DISMISSED** [review]
LGTM

**Comment by jiaqiluo** [comment]
@cbron  Need your final approval then this PR can be merged.

File: packages/rancher-alerting-drivers/charts/templates/hardened.yaml
Comment (cbron) [reply packages/rancher-alerting-drivers/charts/templates/hardened.yaml:25]: Why is the namespace in the arg instead of the command ? Also is specifying the namespace required if the SA is from the same namespace ? I'm not sure on that one, depends how default SA and kubectl act.

File: packages/rancher-alerting-drivers/charts/templates/hardened.yaml
Comment (cbron) [inline packages/rancher-alerting-drivers/charts/templates/hardened.yaml:8]: Should this be pre-install/upgrade, as the default SA already exists. That way it gets edited before the custom SA gets installed.

File: packages/rancher-alerting-drivers/charts/templates/hardened.yaml
Comment (cbron) [inline packages/rancher-alerting-drivers/charts/templates/hardened.yaml:36]: In same vein as above comment, why are these in hooks, can't they be installed at same time as everything else ?

**Review by cbron — COMMENTED** [review]
Logic is fine, just some questions on timing.

File: packages/rancher-alerting-drivers/charts/templates/hardened.yaml
Comment (jiaqiluo) [reply packages/rancher-alerting-drivers/charts/templates/hardened.yaml:8]: The edit on the default SA can happen at either pre or post of the install/upgrade. 
I can move it to pre-install/upgrade if that makes the logic more clear.

File: packages/rancher-alerting-drivers/charts/templates/hardened.yaml
Comment (jiaqiluo) [reply packages/rancher-alerting-drivers/charts/templates/hardened.yaml:36]: We put them in the hooks because editing the default SA is a one-time operation so there is no need to keep those resources (job, ServiceAccount, ClusterRole, ClusterRoleBinding) in the cluster after the operation is done.

File: packages/rancher-alerting-drivers/charts/templates/hardened.yaml
Comment (jiaqiluo) [reply packages/rancher-alerting-drivers/charts/templates/hardened.yaml:25]: The job template is from https://github.com/rancher/charts/blob/dev-v2.5-source/packages/rancher-cis-benchmark/charts/templates/patch_default_serviceaccount.yaml

I agree that it is not necessary to pass the ns as an arg. I will update the PR.

**Review by aiyengar2 — APPROVED** [review]

File: packages/rancher-alerting-drivers/charts/templates/hardened.yaml
Comment (aiyengar2) [reply packages/rancher-alerting-drivers/charts/templates/hardened.yaml:8]: It's post-install since Monitoring V2 needed this job to run after the Namespace resource was created by Helm. It's not a requirement for Alerting Drivers tho, since Alerting Drivers doesn't create namespaced Helm resources.

We could change it, but IMO we should just keep it the same if possible.

File: packages/rancher-alerting-drivers/charts/templates/hardened.yaml
Comment (cbron) [reply packages/rancher-alerting-drivers/charts/templates/hardened.yaml:25]: ok lets go with passing explicitly in the command

File: packages/rancher-alerting-drivers/charts/templates/hardened.yaml
Comment (cbron) [reply packages/rancher-alerting-drivers/charts/templates/hardened.yaml:36]: Ahh, that makes sense. Then its only a question of pre vs post.

File: packages/rancher-alerting-drivers/charts/templates/hardened.yaml
Comment (cbron) [reply packages/rancher-alerting-drivers/charts/templates/hardened.yaml:8]: I'm ok with post, on one stipulation. I don't want to run into a race with hardened cluster where the default SA is attempted to be used before it gets patched away.

Comment (jiaqiluo) [reply packages/rancher-alerting-drivers/charts/templates/hardened.yaml:8]: The namespace is created either by Rancher UI or the user prior to installing the chart, so we do not know if the default SA is used by others or not.  What we can confirm is that the rancher-alerting-driver app does not use the default SA. 

As to the case where the default SA is attempted to be used during the period between starting installing the rancher-alerting-drivers app and executing the post hook, how is it different from the default SA is attempted to be used at any time prior to installing the app? IMO, we do not need to worry about it, or that is something the system, not this app, needs to handle.

File: packages/rancher-alerting-drivers/charts/templates/hardened.yaml
Comment (jiaqiluo) [reply packages/rancher-alerting-drivers/charts/templates/hardened.yaml]: Fixed: the arg values were moved to the command.

File: packages/rancher-alerting-drivers/charts/templates/hardened.yaml
Comment (cbron) [reply packages/rancher-alerting-drivers/charts/templates/hardened.yaml:8]: Oh, I was thinking about this wrong. The default SA is only applied if the app doesn't specify a service account. As long as it does the default SA won't be used.

**Review by cbron — APPROVED** [review]

**Review by aiyengar2 — CHANGES_REQUESTED** [review]
This PR needs to be moved to the dev-v2.5 branch as the dev-v2.5-source branch is now deprecated.

The easiest way to do this would be to create a new local branch from the `dev-v2.5` branch, cherry-pick your commits from the branch this PR points to to that new branch, and then run `make charts` to generate the charts/ and assets/ as part of the new provisioning workflow.

More details on the new workflow can be found in https://github.com/rancher/charts/blob/dev-v2.5/packages/README.md#common-workflow. Feel free to find me on Slack / RocketChat or email me at arvind.iyengar@suse.com for any questions.

**Comment by jiaqiluo** [comment]
close in favor of https://github.com/rancher/charts/pull/1216

**Pull request closed without merging**
